### ISSUE TITLE
Crash compiling ISPC file

### ISSUE THREAD

**Issue by kjeffery** [ispc/ispc#1164]
When trying to compile the attached file with 1.8.2 I get the following message:

> 
xlimepony [no job set:/tmp] 121% /rel/third_party/intelispc/1.8.2/ispc Array2D.ispc --target=AVX -o Array2D.o
ispc: /export/users/dybaboki/llvm/llvm-3.6/lib/IR/Instructions.cpp:1669: void llvm::InsertValueInst::init(llvm::Value *, llvm::Value *, ArrayRef<unsigned int>, const llvm::Twine &): Assertion `ExtractValueInst::getIndexedType(Agg->getType(), Idxs) == Val->getType() && "Inserted value must match indexed type!"' failed.
main.cpp(232): FATAL ERROR: Unhandled signal sent to process; terminating.
***
*** Please file a bug report at https://github.com/ispc/ispc/issues
*** (Including as much information as you can about how to reproduce this error).
*** You have apparently encountered a bug in the compiler that we'd like to fix!
***
Abort (core dumped)

Boo! I had to change the extension to txt so that Gibhub would accept it.

[Array2D.ispc.txt](https://github.com/ispc/ispc/files/116758/Array2D.ispc.txt)

**Comment by dbabokin** [comment]
It's reproduced with ispc trunk + LLVM 3.8.

Thanks for the report.

**Comment by kjeffery** [comment]
Thanks, Dmitry!

Brian Green reduced it even further:

```
struct Foo
{
    uniform int f;
};

inline varying Foo func(uniform Foo * varying f)
{
    return *f;
}
```

**Comment by dbabokin** [comment]
Note that while it's legal construction, the result it not well defined and possibly leads to loss of data.

I.e. "uniform Foo * varying" if a varying pointer pointing to uniform data. Dereferencing this pointer results in varying Foo constructed out of number of uniform instances of Foo. This is typically well defined operation. Unless Foo is a struct with "bound uniform" member, i.e. it's uniform in both varying and uniform instances of Foo. As long as there's no natural way to reduce varying to uniform, this operation is not well defined. We'll define it as "get one of the elements of Foo, which f points to", but if "f" points to Foos with different values, only one of them will survive.

**Comment by dbabokin** [comment]
Fixed by #1166.
